[DM-33153] Add Kubernetes support code (#59)
lsst-sqre/safir: docs/kubernetes.rst

########################
Using the Kubernetes API
########################

> Safir clients are encouraged to use the `kubernetes-asyncio <https://github.com/tomplus/kubernetes_asyncio>`__ Python module.

Reviewer: ```suggestion
Safir-based applications are encouraged to use the `kubernetes-asyncio <https://github.com/tomplus/kubernetes_asyncio>`__ Python module.
```

It provides an async API for Kubernetes that will work naturally with FastAPI applications.

Most Kubernetes work can be done by calling that API directly, with no need for Safir wrapper functions.

> Safir provides a convenient :py:func:`~safir.kubernetes.initialize_kubernetes` function that chooses the correct way to load the Kubernetes configuration depending on whether the code is running from within or outside of a cluster, and a framework for mocking the Kubernetes API for tests.

Reviewer: ```suggestion
Safir provides a convenient `~safir.kubernetes.initialize_kubernetes` function that chooses the correct way to load the Kubernetes configuration depending on whether the code is running from within or outside of a cluster, and a framework for mocking the Kubernetes API for tests.
```

Wherever you've got `:py:*:` in the docs, I recommend just dropping the explicit role and instead using single backticks. In Safir's doc's `conf.py` we've got `default_role = "py:obj"`, which makes single backticks using the Python role.

Kubernetes support in Safir is optional.
To use it, depend on ``safir[kuberentes]``.

Initializing Kubernetes
=======================

A Kubernetes configuration must be loaded before making the first API call.
Safir provides the :py:func:`~safir.kubernetes.initialize_kubernetes` async function to do this.
It doesn't take any arguments.
The Kubernetes configuration will be loaded from the in-cluster configuration path if the environment variable ``KUBERNETES_PORT`` is set, which will be set inside a cluster, and otherwise attempts to load configuration from the user's home directory.

A FastAPI application that uses Kubernetes from inside route handlers should normally call this function during application startup.
For example:

.. code-block:: python

   from safir.kubernetes import initialize_kubernetes

   @app.on_event("startup")
   async def startup_event() -> None:
       await initialize_kubernetes()

Testing with mock Kubernetes
============================

> The :py:mod:`safir.testing.kubernetes` module provides a mock Kubernetes API with a (very) partial implementation of the full API, and some utility functions to use it.

Reviewer: ```suggestion
The `safir.testing.kubernetes` module provides a mock Kubernetes API with a limited implementation the API, and some utility functions to use it.
```

Applications that want to run tests with the mock Kubernetes API should define a fixture (in ``conftest.py``) as follows:

.. code-block:: python

   from typing import Iterator

   import pytest

   from safir.testing.kubernetes import MockKubernetesApi, patch_kubernetes

   @pytest.fixture
   def mock_kubernetes() -> Iterator[MockKubernetesApi]:
       yield from patch_kubernetes()

Then, when initializing Kubernetes, be sure not to import ``ApiClient``, ``CoreV1Api``, or ``CustomObjectsApi`` directly into a module.
Instead, use:

.. code-block:: python

   from kubernetes_asyncio import client

and then use ``client.ApiClient``, ``client.CoreV1Api``, and ``client.CustomObjectsApi``.
This will ensure that the Kubernetes API is mocked properly.

You can then use ``mock_kubernetes`` as a fixture.
The resulting object supports a limited subset of the ``client.CoreV1Api`` and ``client.CustomObjectsApi`` method calls for creating, retrieving, modifying, and deleting objects.
The objects created by either the test or by the application code under test will be stored in memory inside the ``mock_kubernetes`` object.

> The method :py:meth:`~safir.testing.kubernetes.MockKubernetesApi.get_all_objects_for_test` can be used to retrieve all objects of a given kind, allowing comparisons against an expected list of objects.

Reviewer: ```suggestion
You can use the `~safir.testing.kubernetes.MockKubernetesApi.get_all_objects_for_test` method to retrieve all objects of a given kind, allowing comparisons against an expected list of objects.
```

(active voice)

Testing error handling
----------------------

The ``mock_kubernetes`` fixture supports error injection by setting the ``error_callback`` attribute on the object to a callable.
If this is set, that callable will be called at the start of every mocked Kubernetes API call.
It will receive the method name as its first argument and the arguments to the method as its subsequent arguments.

> Inside that callable, the test may, for example, make assertions about the arguments passed in to that method or raise exceptions to simulate errors from the Kubernetes API.

Reviewer: It'd be useful to have a simple example, if you can pull one from gafaelfawr or moneypenny.
